lib/puppet/provider/package/puppetserver_gem.rb

```diff
@@ -50,6 +57,60 @@ def self.gemlist(options)
     end
   end
 
+  # The puppetserver gem cli commands are particularly slow as they start a JVM.
+  # Instead, for the often called list command, use the rubygems library from the puppet ruby,
+  #   setting the GEM_HOME and GEM_PATH to the values in the puppetserver configuration file.
+  # The rubygems library does not have access to java platform gems, for example: json (1.8.3 java),
+  #   but java platform gems should not be managed, by design.
+
+  def self.gem_list_local_via_rubygems(gem_regex)
+    pe_puppetserver_conf_file = '/etc/puppetlabs/puppetserver/conf.d/pe-puppet-server.conf'
+    os_puppetserver_conf_file = '/etc/puppetlabs/puppetserver/puppetserver.conf'
+    puppetserver_gem_home = '/opt/puppetlabs/server/data/puppetserver/jruby-gems'
+    puppetserver_gem_path = ['/opt/puppetlabs/server/data/puppetserver/jruby-gems', '/opt/puppetlabs/server/data/puppetserver/vendored-jruby-gems']
+
+    gem_env = {}
+
+    if File.file?(pe_puppetserver_conf_file)
+      puppetserver_conf_file = pe_puppetserver_conf_file
+    else
+      puppetserver_conf_file = os_puppetserver_conf_file
+    end
+
+    puppetserver_conf = Hocon.load(puppetserver_conf_file)
+
+    if puppetserver_conf.empty?
+      gem_env['GEM_HOME'] = puppetserver_gem_home
+      gem_env['GEM_PATH'] = puppetserver_gem_path
+    else
+      gem_env['GEM_HOME'] = puppetserver_conf['jruby-puppet']['gem-home']
+      gem_env['GEM_PATH'] = puppetserver_conf['jruby-puppet']['gem-path'].join(':')
+    end
+
+    Gem.paths = gem_env
```

Comment:
Seems like we should save the current value of this and restore it in an `ensure` block at the end of the method. Otherwise, it seems possible that changing `Gem.paths` could affect the loading of gems later on in the agent run.

Reply:
Maybe something like:

```ruby
def self.gem_list_local_via_rubygems(gem_regex)
  saved_paths = Gem.paths
  ...
ensure
  Gem.paths = saved_paths
end
```

Reply:
Done, using `Gem.clear_paths` which is documented as allowing paths to be recalculated the next time they are requested. Here is the output of inspecting Gem.paths before and after:

```
#<Gem::PathSupport:0x0000000002eef7f8 @home="/opt/puppetlabs/puppet/lib/ruby/gems/2.4.0", @path=["/root/.gem/ruby/2.4.0", "/opt/puppetlabs/puppet/lib/ruby/gems/2.4.0", "/opt/puppetlabs/puppet/lib/ruby/vendor_gems"], @spec_cache_dir="/root/.gem/specs">
#<Gem::PathSupport:0x0000000003062130 @home="/opt/puppetlabs/server/data/puppetserver/jruby-gems", @path=["/opt/puppetlabs/server/data/puppetserver/jruby-gems", "/opt/puppetlabs/server/data/puppetserver/vendored-jruby-gems"], @spec_cache_dir="/dev/null">
#<Gem::PathSupport:0x0000000002ca59e8 @home="/opt/puppetlabs/puppet/lib/ruby/gems/2.4.0", @path=["/root/.gem/ruby/2.4.0", "/opt/puppetlabs/puppet/lib/ruby/gems/2.4.0", "/opt/puppetlabs/puppet/lib/ruby/vendor_gems"], @spec_cache_dir="/root/.gem/specs">
```